File: 0075-sort-colors/0075-sort-colors.java
class Solution {
    public void sortColors(int[] nums) {
        int i = 0;
        while (i < nums.length - 1) {
            int j = i + 1;
            if (nums[i] > nums[j]) {
                swap(nums, i, j);
                if (i == 0) {
                    //dont check if we reached far left
                } else {
                    int k = i - 1;
                    checkBackward(nums, k);
                }
                i++; 
            }
            else if (nums[i] == nums[j]) {
                i++;
            }
            else {
                i++;
            }
        }
    }

    private void checkBackward(int[] nums, int k) {
        if (k < 0) return;
        
        int next = k + 1;
        if (nums[k] > nums[next]) {
            swap(nums, k, next);
            checkBackward(nums, k - 1);
        }
    }

    public void swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
}

File: 0075-sort-colors/0075-sort-colors.py
class Solution(object):
    def sortColors(self, nums):
        """
        :type nums: List[int]
        :rtype: None
        """
        # Pointers for the current element, the boundary of 0s, and the boundary of 2s
        current = 0
        p0 = 0
        p2 = len(nums) - 1
        
        # Traverse through the list
        while current <= p2:
            if nums[current] == 0:
                nums[current], nums[p0] = nums[p0], nums[current]
                p0 += 1
                current += 1
            elif nums[current] == 2:
                nums[current], nums[p2] = nums[p2], nums[current]
                p2 -= 1
            else:
                current += 1

